### PR TITLE
fee transfer to treasury

### PR DESCRIPTION
**Pull request by akhilesh-singh-Lab49**
fee transfer to treasury account.

### CODE REVIEW

File: contracts/AbstractPair.sol [hunk 1]
```diff
@@ -106,18 +110,23 @@ abstract contract AbstractPair is IPair, HederaResponseCodes {
...
-        _deltaAQty -= feeForToken(_deltaAQty);
+        int feeTokenA = feeForToken(_deltaAQty);
+        _deltaAQty = _deltaAQty - (feeTokenA/2);
+        
         int deltaBQty = getOutGivenIn(_deltaAQty);
         pair.tokenA.tokenQty += _deltaAQty;  
+        int feeTokenB = feeForToken(deltaBQty);
         // deduct fee from the token B
-        deltaBQty -= feeForToken(deltaBQty);
+        deltaBQty = deltaBQty - (feeTokenB/2);
```
Comment (amandeeprehal-lab49) [inline contracts/AbstractPair.sol:120]: you can make it consistent like you are doing below.

File: contracts/AbstractPair.sol [hunk 2]
```diff
@@ -240,4 +253,11 @@ abstract contract AbstractPair is IPair, HederaResponseCodes {
         return tokenQ;
     }
 
+    function transferFeeToTreasuary(int feeTokenA, int feeTokenB, address _treasure) private {
```
Comment (amandeeprehal-lab49) [inline contracts/AbstractPair.sol:256]: Why are you passing _treasure? You should use -

`address private treasury;`

As for a contract treasurer should be fixed for its lifetime. There is no check on `transferFeeToTreasuary` so anyone can collect fee. It looks security gap to me.

File: contracts/AbstractPair.sol [hunk 3]
```diff
@@ -240,4 +253,11 @@ abstract contract AbstractPair is IPair, HederaResponseCodes {
...
+        int response = transferToken(pair.tokenB.tokenAddress, address(this), _treasure, feeTokenA);
```
Comment (amandeeprehal-lab49) [inline contracts/AbstractPair.sol:257]: You are passing wrong arguments. You are transferring tokenB but fee is of A feeTokenA.

Also double check the require's message for token

File: contracts/AbstractPair.sol [hunk 4]
```diff
@@ -240,4 +253,11 @@ abstract contract AbstractPair is IPair, HederaResponseCodes {
...
+        require(response == HederaResponseCodes.SUCCESS, "swapFeeTokenA: Transferring token A to treasuary failed with status code");
+
+        response = transferToken(pair.tokenB.tokenAddress, address(this), _treasure, feeTokenB);
+        require(response == HederaResponseCodes.SUCCESS, "swapFeeTokenB: Transferring token B to treasuary failed with status code");
```
Comment (amandeeprehal-lab49) [inline contracts/AbstractPair.sol:261]: Same as above

File: contracts/AbstractPair.sol (hunk 3 above)
Comment (akhilesh-singh-Lab49) [reply contracts/AbstractPair.sol:257]: Done

File: contracts/AbstractPair.sol (hunk 4 above)
Comment (akhilesh-singh-Lab49) [reply contracts/AbstractPair.sol:261]: Done

File: contracts/AbstractPair.sol (hunk 2 above)
Comment (akhilesh-singh-Lab49) [reply contracts/AbstractPair.sol:256]: I did that to test that individual function but as it is private and we are capturing treasury in initialize function hence not required.
Removed.

File: contracts/AbstractPair.sol [hunk 5]
```diff
@@ -106,18 +110,23 @@ abstract contract AbstractPair is IPair, HederaResponseCodes {
         int localSlippage = getSlippage();
         require(calculatedSlippage <= (localSlippage),  "Slippage threshold breached.");
         // deduct fee from the token A
-        _deltaAQty -= feeForToken(_deltaAQty);
+        int feeTokenA = feeForToken(_deltaAQty);
+        _deltaAQty = _deltaAQty - (feeTokenA/2);
```
[the remark this answers is not in the document]
Comment (akhilesh-singh-Lab49) [reply contracts/AbstractPair.sol:114]: Token quantity test is already covered in "Swap 1 units of token A  " & "Swap 1 units of token B  "

File: package-lock.json [hunk 6]
```diff
@@ -35993,7 +35993,9 @@
           "resolved": "https://registry.npmjs.org/@typechain/ethers-v5/-/ethers-v5-2.0.0.tgz",
           "integrity": "sha512-0xdCkyGOzdqh4h5JSf+zoWx85IusEjDcPIwNEHP8mrWSnCae4rvrqB+/gtpdNfX7zjlFlZiMeePn2r63EI3Lrw==",
           "dev": true,
-          "requires": {}
+          "requires": {
```
Comment (amandeeprehal-lab49) [inline package-lock.json:35996]: Why this change? This is not required.

**Review by amandeeprehal-lab49 — APPROVED** [review]
